### PR TITLE
[WIP] Tree shaking

### PR DESCRIPTION
**Pull request by twokul**
Ran against a brand new ember application.

Relevant repos:

https://github.com/twokul/broccoli-evict
https://github.com/twokul/assembler

Before:

```sh
➜  tree-shake-test git:(master) ✗ ember build --environment=production
cleaning up...
Built project successfully. Stored in "dist/".
File sizes:
 - dist/assets/tree-shake-test-d41d8cd98f00b204e9800998ecf8427e.css: 0 B
 - dist/assets/tree-shake-test-dd4bab0a373ab16e74ba3ab396062ee3.js: 5.22 KB (1.36 KB gzipped)
 - dist/assets/vendor-0d61d4a0db360286ba05f371e537db50.js: 737.37 KB (191.09 KB gzipped)
 - dist/assets/vendor-d41d8cd98f00b204e9800998ecf8427e.css: 0 B
```

After:

```sh
➜  tree-shake-test git:(master) ✗ ember build --environment=production
cleaning up...
Built project successfully. Stored in "dist/".
File sizes:
 - dist/assets/tree-shake-test-d41d8cd98f00b204e9800998ecf8427e.css: 0 B
 - dist/assets/tree-shake-test-dd4bab0a373ab16e74ba3ab396062ee3.js: 5.22 KB (1.36 KB gzipped)
 - dist/assets/vendor-aff4a77e7369abcdead942027854e865.js: 732.69 KB (190.33 KB gzipped)
 - dist/assets/vendor-d41d8cd98f00b204e9800998ecf8427e.css: 0 B
```

### CODE REVIEW

File: lib/broccoli/bundler.js [hunk 1]
```diff
@@ -149,7 +152,24 @@ module.exports = class Bundler {
 
     let scriptOutputFiles = options.scriptOutputFiles;
 
-    let appJs = concat(applicationAndDepsTree, {
+    let name = this.name;
+    let tree = applicationAndDepsTree;
+    if (experiments.TREE_SHAKING) {
+      tree = new BroccoliEvict(applicationAndDepsTree, {
```
Comment (twokul) [inline lib/broccoli/bundler.js]: per conversation with @rwjblue, we should likely create `broccoli-assembler` & move `new Assembler(...)` & `concat` code there.

File: lib/broccoli/bundler.js [hunk 2]
```diff
@@ -149,7 +152,24 @@ module.exports = class Bundler {
 
     let scriptOutputFiles = options.scriptOutputFiles;
 
-    let appJs = concat(applicationAndDepsTree, {
+    let name = this.name;
+    let tree = applicationAndDepsTree;
+    if (experiments.TREE_SHAKING) {
```
Comment (twokul) [inline lib/broccoli/bundler.js]: `experiments.TREE_SHAKING && process.env.TREE_SHAKING`

File: lib/broccoli/bundler.js (hunk 1 above)
Comment (twokul) [reply lib/broccoli/bundler.js]: make sure to mention timings.

**Comment by rwjblue** [comment]
Closing as I believe this has been surpassed by more recent work, please reopen if this is still being worked on...

**Pull request closed without merging**